sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala

```diff
@@ -0,0 +1,400 @@
+      val inputDF1 = input1.toDF
+        .withColumnRenamed("value", "value1")
+        .withColumn("eventTime1", timestamp_seconds($"value1"))
+        .withWatermark("eventTime1", "0 seconds")
+
+      val input2 = MemoryStream[Int]
+      val inputDF2 = input2.toDF
+        .withColumnRenamed("value", "value2")
+        .withColumn("eventTime2", timestamp_seconds($"value2"))
+        .withWatermark("eventTime2", "0 seconds")
+
+      val stream = inputDF1.join(inputDF2, expr("eventTime1 = eventTime2"), "inner")
+        .groupBy(window($"eventTime1", "5 seconds") as 'window)
+        .agg(count("*") as 'count)
+        .select($"window".getField("start").cast("long").as[Long], $"count".as[Long])
+
+      testStream(stream)(
+        AddData(input1, 1, 2, 3, 4),
+        AddData(input2, 1, 2, 3, 4),
+
+        // op1 W (0, 0)
+        // join output: (1, 1), (2, 2), (3, 3), (4, 4)
+        // state: (1, 1), (2, 2), (3, 3), (4, 4)
+        // op2 W (0, 0)
+        // agg: [0, 5) 4
+        // output: None
+        // state: [0, 5) 4
+
+        // no-data batch triggered
+
+        // op1 W (0, 4)
+        // join output: None
+        // state: (4, 4)
+        // agg: [0, 5) 4
+        // output: None
+        // state: [0, 5) 4
+        CheckNewAnswer(),
+        assertNumStateRows(Seq(1, 0)),
+        assertNumRowsDroppedByWatermark(Seq(0, 0)),
+
+        // Move the watermark
+        AddData(input1, 5),
+        AddData(input2, 5),
+
+        // op1 W (4, 5)
+        // join output: (5, 5)
+        // state: (5, 5)
+        // op2 W (4, 5)
+        // agg: [5, 10) 1
+        // output: (0, 5) 4
+        // state: [5, 10) 1
+
+        // no-data batch triggered
+
+        // op1 W (5, 5)
+        // join output: None
+        // state: (5, 5)
+        // agg: None
+        // output: None
+        // state: [5, 10) 1
+        CheckNewAnswer((0, 4)),
+        assertNumStateRows(Seq(1, 0)),
+        assertNumRowsDroppedByWatermark(Seq(0, 0))
+      )
+    }
+  }
+
+  test("aggregation -> stream deduplication, append mode") {
+    withSQLConf("spark.sql.streaming.unsupportedOperationCheck" -> "false") {
+      val inputData = MemoryStream[Int]
+
+      val aggStream = inputData.toDF()
+        .withColumn("eventTime", timestamp_seconds($"value"))
+        .withWatermark("eventTime", "0 seconds")
+        .groupBy(window($"eventTime", "5 seconds") as 'window)
+        .agg(count("*") as 'count)
+        .withColumn("windowEnd", expr("window.end"))
+
+      // dropDuplicates from aggStream without event time column for dropDuplicates - the
+      // state does not get trimmed due to watermark advancement.
+      val dedupNoEventTime = aggStream
+        .dropDuplicates("count", "windowEnd")
+        .select(
+          $"windowEnd".cast("long").as[Long],
+          $"count".as[Long])
+
+      testStream(dedupNoEventTime)(
+        AddData(inputData, 1, 5, 10, 15),
+        CheckNewAnswer((5, 1), (10, 1), (15, 1)),
```

Comment:
Shall we add the same level of trace comment here for W1 and W2?

Reply:
done